### PR TITLE
DOC-59: Assets JSDoc

### PR DESCRIPTION
**Pull request by ctrlaltdavid**
Case: https://highfidelity.atlassian.net/browse/DOC-59

Note: No executable code changes.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21138/HighFidelity-Beta-PR16133-f384b79.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21138/HighFidelity-Beta-Interface-PR16133-f384b79.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21138/HighFidelity-Beta-PR16133-f384b79.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21138/HighFidelity-Beta-Interface-PR16133-f384b79.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21138/HighFidelity-Beta-PR16133-f384b79.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21138/HighFidelity-Quest-Beta-PR16133-f384b79.apk)

File: libraries/script-engine/src/AssetScriptingInterface.h [hunk 1]
```diff
@@ -25,7 +25,15 @@
...
 
 /**jsdoc
- * The Assets API allows you to communicate with the Asset Browser.
+ * The <code>Assets</code> API provides facilities for interacting with the domain's asset server and the client cache. 
+ * cache.
+ * <p>Assets are stored in the asset server in files with SHA256 names. These files are mapped to user-friendly URLs of the 
+ * format: <code>atp:/path/filename</code>. The assets may optionally be baked, in which case a request for the original 
+ * unbaked version of the asset is automatically redirected to the baked version. The asset data may optionally be stored as 
+ * compressed.</p>
+ * <p>The client cache can be access directly, using <code>"atp:"</code> <code>"cache:"</code> URLs. Interface, avatar, and
```
Comment (jennaingersoll) [inline libraries/script-engine/src/AssetScriptingInterface.h]:
```suggestion
 * <p>The client cache can be access directly, using <code>"atp:"</code> or? <code>"cache:"</code> URLs. Interface, avatar, and 
```

File: libraries/script-engine/src/AssetScriptingInterface.h [hunk 2]
```diff
@@ -41,251 +49,490 @@ class AssetScriptingInterface : public BaseAssetScriptingInterface, QScriptable
...
+     * Called when an {@link Assets.downloadData} call is complete.
+     * @callback Assets~downloadDataCallback
+     * @param {string} data - The content that was downloaded.
+     * @param {Assets.DownloadDataError} error - The success or failure of the download.
      */
     /**jsdoc
-     * Called when downloadData is complete
-     * @callback Assets~downloadDataCallback
-     * @param data {string} content that was downloaded
+     * Downloads content from the asset server, form a SHA256-named file.
```
Comment (jennaingersoll) [inline libraries/script-engine/src/AssetScriptingInterface.h]:
```suggestion
     * Downloads content from the asset server, from a SHA256-named file.
```

File: libraries/script-engine/src/AssetScriptingInterface.h [hunk 3]
```diff
@@ -41,251 +49,490 @@ class AssetScriptingInterface : public BaseAssetScriptingInterface, QScriptable
...
+     * @param {string} url - The raw URL of asset to download: <code>atp:</code> followed by the assets's SHA256 hash.
+     * @param {Assets~downloadDataCallback} callback - The function to call upon completion.
+     * @example <caption>Store and retrieve a string from the asset server.</caption>
+     * var assetURL;
+     * 
+     * // Store the string.
+     * Assets.uploadData("Hello world!", function (url, hash) {
+     *     assetURL = url;
+     *     print("url: " + assetURL);  // atp:a0g89...
+     *     Assets.setMapping("/assetsExamples/hellowWorld.txt", hash, function (error) {
```
Comment (jennaingersoll) [inline libraries/script-engine/src/AssetScriptingInterface.h]:
```suggestion
     *     Assets.setMapping("/assetsExamples/helloWorld.txt", hash, function (error) {
```

File: libraries/script-engine/src/AssetScriptingInterface.h [hunk 4]
```diff
@@ -41,251 +49,490 @@ class AssetScriptingInterface : public BaseAssetScriptingInterface, QScriptable
...
+     * @param {Assets.PutResult} result - Information on the content uploaded.
      */
-
     /**jsdoc
-     * A set of properties that can be passed to {@link Assets.putAsset}.
-     * @typedef {object} Assets.PutOptions
-     * @property {ArrayBuffer|string} [data] byte buffer or string value representing the new asset's content
-     * @property {string} [path=null] ATP path mapping to automatically create (upon successful upload to hash)
-     * @property {boolean} [compress=false] whether to gzip compress data before uploading
+     * Uploads content to the assert server and sets a path-to-hash mapping.
```
Comment (jennaingersoll) [inline libraries/script-engine/src/AssetScriptingInterface.h]:
```suggestion
     * Uploads content to the asset server and sets a path-to-hash mapping.
```

File: libraries/script-engine/src/AssetScriptingInterface.h [hunk 5]
```diff
@@ -25,7 +25,15 @@
 #include <QtNetwork/QNetworkDiskCache>
 
 /**jsdoc
- * The Assets API allows you to communicate with the Asset Browser.
+ * The <code>Assets</code> API provides facilities for interacting with the domain's asset server and the client cache. 
+ * cache.
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply libraries/script-engine/src/AssetScriptingInterface.h]: Fixed.

File: libraries/script-engine/src/AssetScriptingInterface.h (hunk 1 above)
Comment (ctrlaltdavid) [reply libraries/script-engine/src/AssetScriptingInterface.h]: Fixed.

File: libraries/script-engine/src/AssetScriptingInterface.h (hunk 2 above)
Comment (ctrlaltdavid) [reply libraries/script-engine/src/AssetScriptingInterface.h]: Fixed.

File: libraries/script-engine/src/AssetScriptingInterface.h (hunk 3 above)
Comment (ctrlaltdavid) [reply libraries/script-engine/src/AssetScriptingInterface.h]: Fixed.

File: libraries/script-engine/src/AssetScriptingInterface.h (hunk 4 above)
Comment (ctrlaltdavid) [reply libraries/script-engine/src/AssetScriptingInterface.h]: Fixed.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21154/HighFidelity-Beta-PR16133-49ab504.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21154/HighFidelity-Quest-Beta-PR16133-49ab504.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21154/HighFidelity-Beta-PR16133-49ab504.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21154/HighFidelity-Beta-Interface-PR16133-49ab504.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21154/HighFidelity-Beta-PR16133-49ab504.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21154/HighFidelity-Beta-Interface-PR16133-49ab504.exe

**Pull request merged**